File: YOLO8/detect_phone.py
from ultralytics import YOLO
import logging

def find_largest_phone_from_cords(list_cord_points):
    """Finds the largest detected phone (based on bounding box area)."""
    if len(list_cord_points) == 0:
        return []
    return [max(list_cord_points, key=lambda box: (box[2] - box[0]) * (box[3] - box[1]))]


def filter_by_confidence(list_cord_points):
    """Filters detections based on a dynamic confidence threshold."""
    if len(list_cord_points) == 0:
        return []
    return sorted(list_cord_points, key=lambda box: box[4])


def get_phones_from_cords(image, list_cord_points):
    """Crops detected phones from an image based on bounding box coordinates."""
    if len(list_cord_points) == 0:
        return []
    cropped_images = []
    for x1, y1, x2, y2, _ in list_cord_points:
        cropped = image[y1:y2, x1:x2]

        # Ensure the crop is valid before adding
        if cropped.shape[0] > 0 and cropped.shape[1] > 0:
            cropped_images.append(cropped)

    return cropped_images


class DetectPhone:
    min_confidence = 0.3

    def __init__(self, model_path="best.pt"):
        """Initialize the YOLO model for phone detection."""
        self.model = YOLO(model_path)
        logging.getLogger("ultralytics").setLevel(logging.WARNING)

    def find_cord_for_phones(self, image):
        """Detect phones in an image and return their bounding boxes and confidence scores."""
        results = self.model(image, verbose=False)
        phones = []
        for result in results[0].boxes:
            if result.conf > self.min_confidence:
                x1, y1, x2, y2 = map(int, result.xyxy[0])
                phones.append((x1, y1, x2, y2, float(result.conf)))
        return phones

File: Client/src/top-images-and-yolo-at-client/YOLO8/detect_phone.js

/**
 * @typedef {[number, number, number, number, number]} CordsPoint
 * @typedef {Array<Array<number>>} Image
 */

/**
 * Finds the largest phone bounding box based on area.
 * @param {CordsPoint[]} cordPoints
 * @returns {CordsPoint[]}
 */
function findLargestPhoneFromCords(cordPoints) {
  if (cordPoints.length === 0) return [];
  return [
    cordPoints.reduce((largest, current) => {
      const areaCurrent = (current[2] - current[0]) * (current[3] - current[1]);
      const areaLargest = (largest[2] - largest[0]) * (largest[3] - largest[1]);
      return areaCurrent > areaLargest ? current : largest;
    })
  ];
}

/**
 * Sorts bounding boxes by confidence.
 * @param {CordsPoint[]} cordPoints
 * @returns {CordsPoint[]}
 */
function sortByConfidence(cordPoints) {
  return [...cordPoints].sort((a, b) => a[4] - b[4]);
}

/**
 * Crops detected phones from image based on bounding boxes.
 * @param {Image} image - 2D or 3D array representing the image
 * @param {CordsPoint[]} cordPoints
 * @returns {Image[]}
 */
function getPhonesFromCords(image, cordPoints) {
  const croppedImages = [];

  for (const [x1, y1, x2, y2] of cordPoints) {
    const cropped = cropImage(image, x1, y1, x2, y2);
    if (cropped.length > 0 && cropped[0].length > 0) {
      croppedImages.push(cropped);
    }
  }

  return croppedImages;
}

/**
 * Simulates image cropping logic.
 * This function would depend on how you represent image matrices in your project.
 */
function cropImage(image, x1, y1, x2, y2) {
  const cropped = [];
  for (let y = y1; y < y2; y++) {
    if (image[y]) {
      cropped.push(image[y].slice(x1, x2));
    }
  }
  return cropped;
}

// --- Detector class stub ---
class DetectPhone {
  constructor(modelPath = "best.pt") {
    this.MIN_CONFIDENCE = 0.3;
    this.modelPath = modelPath;
    // TODO: Load your model (ONNX.js, TensorFlow.js, etc.)
  }

  /**
   * Simulated function. Replace with actual model inference logic.
   * @param {Image} image
   * @returns {Promise<CordsPoint[]>}
   */
  async findCordForPhones(image) {
    // Example stub. Replace with actual detection logic using your model.
    const mockResults = [
      [30, 40, 130, 200, 0.85],
      [60, 90, 110, 170, 0.60],
    ];
    return mockResults.filter(([,,, , conf]) => conf > this.MIN_CONFIDENCE);
  }
}
